### PR TITLE
#179316675 user should pay for VAS - Part 2

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Update offline payment model to accommodate multiple models

#### Description of Task to be completed?
Update offline payment model to accommodate multiple models

#### How should this be manually tested?
`/vas/request/`

#### What are the relevant pivotal tracker stories?
#179316675

### CODE REVIEW

File: server/services/offlinePayment.service.js [hunk 1]
```diff
@@ -14,19 +14,41 @@ import NOTIFICATIONS from '../helpers/notifications';
...
-export const addOfflinePayment = async (offlinePayment) => {
-  const offer = await getOfferById(offlinePayment.offerId).catch((error) => {
-    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
-  });
+export const verifyPaymentModel = async (offlinePayment) => {
+  let model;
+
+  switch (offlinePayment.modeType) {
+    case VALID_TRANSACTION_MODELS.OFFER:
+      model = await getOfferById(offlinePayment.modelId).catch((error) => {
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:27]:
```js
/// not a valid code, it is a psedo code
const async getModelById = (model, modelId) => {
   switch (model) {
      case MODEL.OFFER:
           return () => await getOfferById(modelId);
      case MODEL.VAS_REQUESTS:
           return () => await getVasRequestById(modelId);

     ....
   }
}
// We should have a model constant

const MODEL = {
   OFFER: 'offers',
   VAS_REQUEST: 'vasrequests',
}
````

File: server/services/offlinePayment.service.js [hunk 2]
```diff
@@ -42,7 +64,7 @@ export const addOfflinePayment = async (offlinePayment) => {
     );
   }
 
-  const property = await getPropertyById(offer.propertyId);
+  const property = await getPropertyById(model.propertyId);
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:67]: Model will not always have propertyId. We should look for a better way to implement this.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. This is a good start but we should focus on offlinePayment for this.

File: server/models/offlinePayment.model.js [hunk 3]
```diff
@@ -77,10 +81,14 @@ const OfflinePaymentSchema = new mongoose.Schema(
       type: Date,
       required: true,
     },
-    offerId: {
+    modelId: {
```
[the remark this answers is not in the document]
Comment (andela-hpopoola) [reply server/models/offlinePayment.model.js:84]: Not sure why you didn't implement this as we have explained here.

File: server/services/helper.service.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { ErrorHandler } from '../helpers/errorHandler';
+import httpStatus from '../helpers/httpStatus';
+import { MODEL } from '../helpers/constants';
+import { getOfferById } from './offer.service';
+import { getVasRequestById } from './vasRequest.service';
+
+// eslint-disable-next-line import/prefer-default-export
+export const verifyModel = async ({ modelType, modelId }) => {
```
Comment (andela-hpopoola) [inline server/services/helper.service.js:8]: Not sure why you used `verifyModel` when we are getting all the models by the Id. 

```suggestion
export const getModelById = async ({ modelType, modelId }) => {
```

File: server/services/offlinePayment.service.js [hunk 5]
```diff
@@ -42,17 +37,26 @@ export const addOfflinePayment = async (offlinePayment) => {
...
     const payment = await new OfflinePayment(offlinePayment).save();
-    const description = `You added an offline payment of ${getMoneyFormat(
-      payment.amount,
-    )} for ${getFormattedName(property.name)}`;
-    await createNotification(NOTIFICATIONS.OFFLINE_PAYMENT_ADDED, offlinePayment.userId, {
-      description,
-      actionId: payment._id,
-    });
+
+    if (model.propertyId) {
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:43]: I think it would be better to check the `modelType` instead of propertyId.

```suggestion
    if (modelType === MODEL.OFFER) {
```

Comment (andela-hpopoola) [reply server/services/offlinePayment.service.js:43]: Also, it is better to generate the description and use just one await `createNotification(...` instead of including it in the `if` and `else` block.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for making the changes. I left a few comments to improve the concept of the implementation. Can you write the tests to confirm that the implementation works as expected.

PS: Can you work on the vendor profile page as we discussed before fixing the comments here. I expect you to have a PR ready now so that we can round it up before the end of today so that we can implement it before sprint planning. Lets make this a priority please 🙏🏽

File: server/schemas/offlinePayment.schema.js [hunk 6]
```diff
@@ -9,12 +9,33 @@ import {
...
+  id: optionalObjectId('Model Id'),
+  type: optionalModelType,
+});
 
 export const addOfflinePaymentSchema = Joi.object({
   amount: requiredNumber('Amount'),
   bankId: requiredObjectId('Bank id'),
   dateOfPayment: requiredDate('Payment Date'),
-  offerId: requiredObjectId('Offer id'),
+  model,
```
Comment (andela-hpopoola) [inline server/schemas/offlinePayment.schema.js:38]: Sometimes, this implementation usually feels difficult to read but I usually overlook them. I believe it would be easier to read if you name the validation with `requiredModel` instead of model.

```suggestion
  model: requiredModel,
```

File: server/services/offlinePayment.service.js [hunk 7]
```diff
@@ -42,17 +39,21 @@ export const addOfflinePayment = async (offlinePayment) => {
...
 
-  const property = await getPropertyById(offer.propertyId);
-
   try {
     const payment = await new OfflinePayment(offlinePayment).save();
-    const description = `You added an offline payment of ${getMoneyFormat(
-      payment.amount,
-    )} for ${getFormattedName(property.name)}`;
+
+    if (offlinePayment.model.type === VALID_TRANSACTION_MODELS.OFFER) {
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:45]: Since we are in the offline payment service, using `VALID_TRANSACTION_MODELS` looks awkward so it is better to use `MODEL`

```suggestion
    if (offlinePayment.model.type === MODEL.OFFER) {
```

File: server/services/offlinePayment.service.js [hunk 8]
```diff
@@ -268,12 +297,16 @@ export const raiseComment = async ({ comment, user }) => {
       },
       { new: true, safe: true, upsert: true },
     );
-    const offer = await getOffer(offlinePayment.offerId);
 
-    const description = `A comment has been raised on your offline payment for ${getFormattedName(
-      offer[0].propertyInfo.name,
-    )}`;
-    await createNotification(NOTIFICATIONS.RAISE_COMMENT_FOR_OFFLINE_PAYMENT, offer[0].userId, {
+    if (offlinePayment.model.type === VALID_TRANSACTION_MODELS.OFFER) {
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:301]: Same comment as above.

File: server/services/offlinePayment.service.js [hunk 9]
```diff
@@ -14,18 +14,15 @@ import NOTIFICATIONS from '../helpers/notifications';
...
 export const addOfflinePayment = async (offlinePayment) => {
-  const offer = await getOfferById(offlinePayment.offerId).catch((error) => {
-    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
-  });
+  let description = `You added an offline payment`;
 
-  if (!offer) {
-    throw new ErrorHandler(httpStatus.NOT_FOUND, 'Invalid offer');
-  }
+  const model = await getModelById(offlinePayment);
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:25]: We need to add tests to confirm that we can add offline payments for vas requests. All the important tests written for offer/property in respect to offline payments should be written for vas requests.

File: server/services/offlinePayment.service.js [hunk 10]
```diff
@@ -237,6 +264,8 @@ export const resolveOfflinePayment = async ({ offlinePaymentId, adminId }) => {
 };
 
 export const raiseComment = async ({ comment, user }) => {
+  let description = `A comment has been raised on your offline payment`;
```
Comment (andela-hpopoola) [inline server/services/offlinePayment.service.js:267]: A test should be written to confirm that this also works as expected for vas request.

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. I left a few more comments on adding more tests to confirm that we have covered most edge cases for vas requests in respect to offline payments. However, this looks good to me and I trust that you should be able to write the necessary tests before merging. Thank you.
